apps/prairielearn/src/lib/group-update.ts

```diff
@@ -192,10 +192,33 @@ export async function randomGroups(
             { assessment_id },
             z.number(),
           );
+
           // Create groups using the groups of maximum size where possible
-          for (let i = unusedGroupNameSuffix ?? 1; studentsToGroup.length > 0; i++) {
+          const userGroups = _.chunk(
+            studentsToGroup.map((user) => user.uid),
+            max_group_size,
+          );
+          // If the last group is too small, move students from larger groups to the last group
+          const smallGroup = userGroups.at(-1);
+          while (smallGroup && smallGroup.length < min_group_size) {
```

Comment:
Checking my understanding: does this handle the case where we have `min=4, max=6` and groups of size `6, 6, 1`? In that case, after one iteration we'd have `5, 5, 3`, and we'd need one more iteration to get to `5, 4, 4`. Is that right?

Reply:
Yes, that's correct. The vast majority of cases will not require more than one iteration of the loop, since in most cases getting one student from each group is enough to fill the last group. A second iteration (or more) is only needed for cases where the minimum number of students in each group is higher than the number of groups overall, since in that case the first iteration may not cause the smaller group to reach the minimum. I don't expect this to be a common case, it is added for completeness.